### ISSUE TITLE
Undeterminable Kinetics Error of Surface_Abstraction_vdW family

### ISSUE THREAD

**Issue by Tingchenlee** [ReactionMechanismGenerator/RMG-Py#2122]
### Bug Description
The reactions crashed when the bidentate species tried to match the reverse template of Surface_Abstraction_vdW family. There were three problem reactions so far:
![image](https://user-images.githubusercontent.com/56702445/117471687-220fa200-af26-11eb-9af8-5e1df87ad0d7.png)

### How To Reproduce
Maybe an easier way to reproduce these is using the **kinetic search** of the RMG website, here are the error messages
 - Reaction 1.
```
rmgpy.exceptions.UndeterminableKineticsError: 
(TemplateReaction(reactants=[Molecule(smiles="NO[O].[Pt]"), Molecule(smiles="[N-]([NH+]=[Pt])[Pt]")], 
products=[Molecule(smiles="[N-]([NH+](O[O])[Pt])[Pt]"), Molecule(smiles="N[Pt]")], 
pairs=[], family='Surface_Abstraction_vdW'), 'Kinetics could not be determined. 
Unable to find matching template for reaction 
<Molecule "NO[O].[Pt]"> + <Molecule "[N-]([NH+]=[Pt])[Pt]"> <=> <Molecule "[N-]([NH+](O[O])[Pt])[Pt]"> + <Molecule "N[Pt]"> 
in reaction family <KineticsGroups "Surface_Abstraction_vdW/groups">.
Trying to match [<Entry index=1 label="AdsorbateVdW">, <Entry index=2 label="Adsorbate1">] but matched [<Entry index=9 label="*N">]')
```
 - Reaction 2.
 ```
rmgpy.exceptions.UndeterminableKineticsError: 
(TemplateReaction(reactants=[Molecule(smiles="NN([Pt])[O].[Pt]"),
Molecule(smiles="O=[Pt]")], products=[Molecule(smiles="[Pt]ON([Pt])[O]"), Molecule(smiles="N[Pt]")], 
pairs=[], family='Surface_Abstraction_vdW'), 'Kinetics could not be determined.
 Unable to find matching template for reaction
 <Molecule "NN([Pt])[O].[Pt]"> + <Molecule "O=[Pt]"> <=> <Molecule "[Pt]ON([Pt])[O]"> + <Molecule "N[Pt]"> 
in reaction family <KineticsGroups "Surface_Abstraction_vdW/groups">.
Trying to match [<Entry index=1 label="AdsorbateVdW">, <Entry index=2 label="Adsorbate1">] but matched [<Entry index=5 label="*=O">]')
```
 - Reaction 3.
 ```
rmgpy.exceptions.UndeterminableKineticsError: 
(TemplateReaction(reactants=[Molecule(smiles="[Pt]ON(N([Pt])[O])[O].[Pt]"), Molecule(smiles="O=[Pt]")], 
products=[Molecule(smiles="[Pt]ON([Pt])[O]"), Molecule(smiles="[Pt]ON([Pt])[O]")], 
pairs=[], family='Surface_Abstraction_vdW'), 'Kinetics could not be determined. 
Unable to find matching template for reaction 
<Molecule "[Pt]ON(N([Pt])[O])[O].[Pt]"> + <Molecule "O=[Pt]"> <=> <Molecule "[Pt]ON([Pt])[O]"> + <Molecule "[Pt]ON([Pt])[O]"> 
in reaction family <KineticsGroups "Surface_Abstraction_vdW/groups">.
Trying to match [<Entry index=1 label="AdsorbateVdW">, <Entry index=2 label="Adsorbate1">] but matched [<Entry index=5 label="*=O">]') 
```

### Expected Behavior
 - Maybe we should forbid the radical and charged species here? (For reaction 1.)
 - Maybe we don't want to have bidentate species react with this family? (For reactions 2. and 3.)
 - Better ways to prevent the crash?

**Comment by Tingchenlee** [comment]
Previous discussion in RMG-website issue https://github.com/ReactionMechanismGenerator/RMG-website/issues/219

**Comment by ChrisBNEU** [comment]
more generally I think it is also related to #2106.

**Comment by davidfarinajr** [comment]
I think PR https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2131 should fix this

**Comment by davidfarinajr** [comment]
I think that PR should fix the `UndeterminableKineticsError`, but it will not fix the multi-dentate issue where multi-dentate things react with reaction families designed for mono-dentate adsorbates.

**Comment by Tingchenlee** [comment]
Similar error in **Surface_Abstraction_Single_vdW** family
```Error: Expecting one matching reverse reaction, not zero in reaction family Surface_Abstraction_Single_vdW for forward reaction [Pt]ON([Pt])[O] + [Pt].N <=> N[Pt] + [O]NO[Pt].[Pt].

Error: There is likely a bug in the RMG-database kinetics reaction family involving a missing group, missing atomlabels, forbidden groups, etc.
Reactant
[Pt]ON([Pt])[O]
multiplicity 2
1 O u0 p2 c0 {3,S} {5,S}
2 O u1 p2 c0 {3,S}
3 N u0 p1 c0 {1,S} {2,S} {4,S}
4 X u0 p0 c0 {3,S}
5 X u0 p0 c0 {1,S}

Reactant
[Pt].N
1 N u0 p1 c0 {2,S} {3,S} {4,S}
2 H u0 p0 c0 {1,S}
3 H u0 p0 c0 {1,S}
4 H u0 p0 c0 {1,S}
5 X u0 p0 c0

Product
N[Pt]
1 N u0 p1 c0 {2,S} {3,S} {4,S}
2 H u0 p0 c0 {1,S}
3 H u0 p0 c0 {1,S}
4 X u0 p0 c0 {1,S}

Product
[O]NO[Pt].[Pt]
multiplicity 2
1 O u0 p2 c0 {3,S} {5,S}
2 O u1 p2 c0 {3,S}
3 N u0 p1 c0 {1,S} {2,S} {4,S}
4 H u0 p0 c0 {3,S}
5 X u0 p0 c0 {1,S}
6 X u0 p0 c0


[O]NO[Pt].[Pt]
multiplicity 2
1 O u0 p2 c0 {3,S} {5,S}
2 O u0 p3 c-1 {3,S}
3 N u1 p0 c+1 {1,S} {2,S} {4,S}
4 H u0 p0 c0 {3,S}
5 X u0 p0 c0 {1,S}
6 X u0 p0 c0

Error: Debugging why no reaction was found...
Error: Checking whether the family's forbidden species have affected reaction generation...
Error: Still experiencing error: Expecting one matching reverse reaction, not 0 in reaction family Surface_Abstraction_Single_vdW for forward reaction [Pt]ON([Pt])[O] + [Pt].N <=> N[Pt] + [O]NO[Pt].[Pt].

rmgpy.exceptions.KineticsError: Did not find reverse reaction in reaction family Surface_Abstraction_Single_vdW for reaction [Pt]ON([Pt])[O] + [Pt].N <=> N[Pt] + [O]NO[Pt].[Pt].
```
A quick solution is applying forbbiden block to avoid bidentates
```
1 *5 X   u0 p0 c0 {2,S}
2 *4 R!H ux px cx {1,S} {3,[S,D]} 
3    R   ux px cx {2,[S,D]} {4,S}
4    X   u0 p0 c0 {3,S}
```

Another error in  **Surface_Dual_Adsorption_vdW** family
```
Error: Problem family: Surface_Dual_Adsorption_vdW
Error: Problem reactants: (Molecule(smiles="[Pt]ON([Pt])O"), Molecule(smiles="[Pt]ON([Pt])[O]"))
Error: <Molecule "[Pt]ON([Pt])O">
1 *6 O u0 p2 c0 {3,S} {6,S}
2    O u0 p2 c0 {3,S} {4,S}
3    N u0 p1 c0 {1,S} {2,S} {5,S}
4    H u0 p0 c0 {2,S}
5    X u0 p0 c0 {3,S}
6 *5 X u0 p0 c0 {1,S}

Error: <Molecule "[Pt]ON([Pt])[O]">
multiplicity 2
1 *2 O u0 p2 c0 {3,S} {5,S}
2 *4 O u1 p2 c0 {3,S}
3 *3 N u0 p1 c0 {1,S} {2,S} {4,S}
4    X u0 p0 c0 {3,S}
5 *1 X u0 p0 c0 {1,S}

rmgpy.exceptions.UndeterminableKineticsError: (TemplateReaction(reactants=[Molecule(smiles="[Pt]ON=O.[Pt]"), Molecule(smiles="[O]NO[Pt].[Pt]")], products=[Molecule(smiles="[Pt]ON([Pt])[O]"), Molecule(smiles="[Pt]ON([Pt])O")], pairs=[], family='Surface_Dual_Adsorption_vdW'), 'Kinetics could not be determined. Unable to find matching template for reaction <Molecule "[Pt]ON=O.[Pt]"> + <Molecule "[O]NO[Pt].[Pt]"> <=> <Molecule "[Pt]ON([Pt])[O]"> + <Molecule "[Pt]ON([Pt])O"> in reaction family <KineticsGroups "Surface_Dual_Adsorption_vdW/groups">.Trying to match [<Entry index=1 label="Adsorbate1">, <Entry index=2 label="Adsorbate2">] but matched [<Entry index=48 label="ONR">]')
```
A quick solution is applying forbbiden block to avoid bidentates
```
1 *5 X   u0 p0 c0 {2,S}
2 *6 R!H ux px cx {1,S} {3,[S,D]} 
3    R   ux px cx {2,[S,D]} {4,S}
4    X   u0 p0 c0 {3,S}
```
and 
```
1 *5 X   u0 p0 c0 {2,S}
2 *6 R!H ux px cx {1,S} {3,[S,D]} 
3    R   ux px cx {2,[S,D]} {4,S}
4    X   u0 p0 c0 {3,S}
```

**Comment by davidfarinajr** [comment]
So I think what is actually causing the `UndeterminableKineticsError` is that the vdw molecules have radicals, and not that they are bidentate.  We recently merged a PR (https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2131), that I think should fix this problem and prevent the families from generating reactions with vdW rads.  Have you tried building a model since that PR was merged?  If so, are these errors still occurring? The vdw bidentate problem could also be causing problems and is an issue we still need to address.

This open PR also globally forbids vdW rads
https://github.com/ReactionMechanismGenerator/RMG-database/pull/481

**Comment by Tingchenlee** [comment]
The errors can be reproduced by running the following input file on vdw_radical branch:
```
database(
    thermoLibraries=['surfaceThermoPt111', 'surfaceThermoNi111', 'primaryThermoLibrary', 'thermo_DFT_CCSDTF12_BAC','DFT_QCI_thermo', 'GRI-Mech3.0-N', 'NitrogenCurran', 'primaryNS', 'CHON'],
    reactionLibraries = ['Surface/CPOX_Pt/Deutschmann2006'], 
    seedMechanisms = [],
    kineticsDepositories = ['training'],
    kineticsFamilies = ['surface','default'],
    kineticsEstimator = 'rate rules',
)

catalystProperties(
    metal = 'Pt111'
)

generatedSpeciesConstraints(
    allowed=['input species','seed mechanisms','reaction libraries'],
    maximumNitrogenAtoms=2,
    maximumOxygenAtoms=3,
)

# List of species
species(
    label='X',
    reactive=True,
    structure=adjacencyList("1 X u0"),
)

species(
    label='O2',
    reactive=True,
    structure=adjacencyList(
"""
multiplicity 3
1 O u1 p2 c0 {2,S}
2 O u1 p2 c0 {1,S}
"""),
)

species(
    label='NH3',
    reactive=True,
    structure=adjacencyList(
"""
1 N u0 p1 c0 {2,S} {3,S} {4,S}
2 H u0 p0 c0 {1,S}
3 H u0 p0 c0 {1,S}
4 H u0 p0 c0 {1,S}
"""),
)

species(
    label='He',
    reactive=False,
    structure=adjacencyList(
"""
1 He u0 p1 c0
"""),
)

surfaceReactor(  
    temperature=(1000,'K'),
    initialPressure=(1.0, 'bar'),
    nSims=12,
    initialGasMoleFractions={
        "NH3": 0.12,
        "O2": 0.88,
        "He": 0.38,
    },
    initialSurfaceCoverages={
        "X": 1.0,
    },
    surfaceVolumeRatio=(2.8571428e3, 'm^-1'), 
    terminationConversion = {"NH3":0.99,},
    #terminationTime=(10, 's'),
)

simulator( 
    atol=1e-18, 
    rtol=1e-12, 
)

model( 
    toleranceKeepInEdge=0.001, 
    toleranceMoveToCore=0.01, 
    toleranceInterruptSimulation=1e8, 
    maximumEdgeSpecies=50000, 
    minCoreSizeForPrune=50, 
    #toleranceThermoKeepSpeciesInEdge=0.5, 
    minSpeciesExistIterationsForPrune=2, 
)

options(
    units='si',
    saveRestartPeriod=None,
    generateOutputHTML=False,
    generatePlots=False, 
    saveEdgeSpecies=False,
    saveSimulationProfiles=False,
)
```

**Comment by davidfarinajr** [comment]
Thanks @Tingchenlee ! I was able to reproduce that error with that input file and those branches.  
We have a lot going on here 🤦 .  There are 3 main issues:
1) **vdw radicals** - this is actually causing RMG to crash with kinetics error.  We implemented a multiplicity check to catch this for products in the reverse direction.  However, we are getting vdw products in the *forward* direction here, so we also need to check this direction as well.  We added `vdw_radical` forbidden struct, but we need to add this *to each vdw family* because I don't think the *global* forbidden structs are checked until after reaction is generated (and we need to check before because it crashes)
2) **radicals on X** - when we fix vdw radical problem, the job continues on and crashes when resonance generates a species with radicals on X.  This commit fixes that ddb17664beef4612d37a0cf861e2fd6eb92026a3
3) bidentate vdw - the first 2 issues cause RMG to crash, not sure if this does but it is an issue we still need to solve


I can open a PR, hopefully tomorrow, to fix 1 and 2 and perhaps 3 but I think more discussion is needed for 3.
